vscripts/confogl.nut

```diff
@@ -2,65 +2,308 @@
 MutationOptions <-
 {
   cm_ProhibitBosses = 0
- 	cm_AllowPillConversion = 0
+  cm_AllowPillConversion = 0
 }
 
-
 local pluginScopes = []
-Hooks <- {
-  function AllowBash(basher, bashee) {
-    // Default to allow the bash.
-    local retVal = true;
-	  for ( local idx = pluginScopes.len(); idx >= 0; --idx ) {
-      local nextPlugin = pluginScopes[idx];
-      local fn = pluginScopes[idx].AllowBash;
-      if ( typeof (fn) == "function") {
-        local res = fn(basher, bashee)
-        // If any plugin returns false, we return false.
-        if(res == false) {
-          return false;
-        }
-      }
+
+// int AllowBash(handle basher, handle bashee): called whenever melee bash is used.
+// Returns these values:
+// ALLOW_BASH_ALL - normal melee behavior
+// ALLOW_BASH_NONE - do nothing at all
+// ALLOW_BASH_PUSHONLY - applies physics push but deals no damage (including prevention of insta-kill ambush behavior)
+function AllowBash(basher, bashee) {
+  Msg(format("AllowBash(%s, %s)\n", basher.tostring(), bashee.tostring()));
+  // Default to allow the bash.
+  local retVal = ALLOW_BASH_ALL;
+  for ( local idx = pluginScopes.len() - 1; idx >= 0; --idx ) {
+    local nextPlugin = pluginScopes[idx];
+    if ("AllowBash" in pluginScopes[idx]) {
+	  local fn = pluginScopes[idx].AllowBash;
+	  if ( typeof (fn) == "function" ) {
+	    local res = fn(basher, bashee)
+	    // If any plugin returns ALLOW_BASH_NONE, we return ALLOW_BASH_NONE.
+	    if(res == ALLOW_BASH_NONE) {
+		  return ALLOW_BASH_NONE;
+	    }
+	    // If any plugin returns ALLOW_BASH_PUSHONLY, we return ALLOW_BASH_PUSHONLY.
+	    if(res == ALLOW_BASH_PUSHONLY) {
+		  return ALLOW_BASH_PUSHONLY;
+	    }
+	  }
+    }
+  }
+  return retVal;
+}
+// bool AllowTakeDamage(table damageTable): Adding a function with this name in the script causes C++ to call it on all damage events.
+// The damageTable is actually defined in scriptedmode.nut and filled in as appropriate before each call.
+function AllowTakeDamage(damageTable) {
+  Msg(format("AllowTakeDamage(%s)\n", damageTable.tostring()));
+  // Default to allow the damage.
+  local retVal = true;
+  for ( local idx = pluginScopes.len() - 1; idx >= 0; --idx ) {
+    local nextPlugin = pluginScopes[idx];
+    if ("AllowTakeDamage" in pluginScopes[idx]) {
+	  local fn = pluginScopes[idx].AllowTakeDamage;
+	  if ( typeof (fn) == "function" ) {
+	    local res = fn(damageTable)
+	    // If any plugin returns false, we return false.
+	    if(res == false) {
+		  return false;
+	    }
+	  }
+    }
+  }
+  return retVal;
+}
```

Comment:
In a case like this, where most/all of the work is being done by side effects, it'd be the first plugin loaded that gets to decide the final mutation of slots in damageTable (since it's last in iteration). This seems undesirable. How should we decide what the final value of each slot should be?